### PR TITLE
Add support for static gzipping

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
Static gzipping is a technique to reduce server load and to allow a higher compression level. All assets should now get served with `gzip -9` level compression, which wasn't realistic if compressed on the fly (hundreds of milliseconds), but made possible by compressing them at build time.

Todo: docs (can someone guide me?)

### CODE REVIEW

**Comment by ishitatsuyuki** [comment]
Rebased. @pi0 Can you review?

**Comment by manniL** [comment]
There are still lint issues @ishitatsuyuki

**Comment by ishitatsuyuki** [comment]
@manniL That's red on dev (target branch), not my fault.

**Comment by manniL** [comment]
Ah, sorry @ishitatsuyuki :see_no_evil:

**Comment by ishitatsuyuki** [comment]
@clarkdo Can you review?

**Comment by ishitatsuyuki** [comment]
@Atinux Can you assign someone?

**Comment by Atinux** [comment]
Hi @ishitatsuyuki 

Sorry for the delay, we have to talk about this with the core team, we also have `render.gzip` so we don't want to keep both options and try to find an elegant solution for gzip support.

**Comment by jetrii** [comment]
I feel like static gzipping should be done with nginx or some reverse proxy. Based on my understanding it would be way faster than doing it in Node.

**Comment by ishitatsuyuki** [comment]
@jetrii Either way, producing a compressed artefact beforehand helps sending maximum compressed bundles. Also, note that to achieve static compression on nginx side you need proxy_cache which takes some effort to set up.

**Comment by ishitatsuyuki** [comment]
@Atinux can you provide an update on this? Sure this may need a discussion, but please avoid postponing it indefinitely.

**Comment by ishitatsuyuki** [comment]
Pinging the full team: @Atinux @pi0 @clarkdo @alexchopin. Please make the decision if necessary, or provide some plan at least.

This is ready for merge. CI failure is due to certificate expiration on https://nuxtjs.org.

**Comment by ishitatsuyuki** [comment]
@Atinux 

> we do believe this should not be in the core of Nuxt but a module instead

I understand that this is achievable with modules, but please explain why you believe this should not be in the Nuxt core.

First, this kind of functionality can be too trivial to be a module. I mean, there's no direct demand of such a small optimisation, and installing a module is an opt-in process here. On the other hand, if this is integrated into Nuxt core, we'll be able to provide (slightly) better performance out of box. Basically, this is the same reason as things like CSP or HTTP/2 push headers lives in the core.

Second, modules are flexible but we're likely going to have some hacks that may break as the Nuxt internals change. For instance, we'll be overriding `serveStatic` by mounting `gzipStatic` before (which has a higher precedence). Doing so can result in bypassing SSR resources check:

https://github.com/ishitatsuyuki/nuxt.js/blob/0d4dd349fd4f274d5d63cd19754b307e49a51f06/lib/core/renderer.js#L222-L230

Also, we need to rely on internal assumptions. If we put such code in a out-of-tree module, it causes maintenance burden.

https://github.com/ishitatsuyuki/nuxt.js/blob/0d4dd349fd4f274d5d63cd19754b307e49a51f06/lib/builder/webpack/client.js#L88

So, I believe modules are not how we should be doing this optimization.

**Comment by ishitatsuyuki** [comment]
bump @Atinux

**Comment by galvez** [comment]
> Second, modules are flexible but we're likely going to have some hacks that may break as the Nuxt internals change.

@ishitatsuyuki this will cease to be a problem in Nuxt V3, as the core will change to accommodate exactly this kind of need painlessly. If the main static serving becomes a module on itself, then it won't be any problem replacing it. 

As insignificant as you may think this is for a module, there's value in keeping the core minimal, even if that means having small and specialized auxiliary packages down the road. Having separate, loosely coupled components will improve the developer experience IMHO.

@Atinux @clarkdo can we close this PR and tag it for the v3 modules?

**Comment by Atinux** [comment]
Indeed, the core will be such minimal that creating other static modules will be easy and answer specific use cases.

I’ll note this PR for having a specific module for it as example of possibilities.

Thank you

**Comment by ishitatsuyuki** [comment]
@galvez

Why would you talk about v3 where there's entirely no information about what's going to change?

> If the main static serving function becomes a module itself, then it won't be any problem replacing it.

Would it? I don't see it being done without a complete rewrite.

Also, both of you discarded my first point of the functionality being trivial and low demand. Why would you not include this in the core? You stated that "you talked with the team" without making the argument clear.

@Atinux Also, please do not leave the discussion stalled for a such long period. That's a extremely unfriendly way to deal with contributors.

**Comment by galvez** [comment]
> Why would you talk about v3 where there's entirely no information about what's going to change?
> Would it? I don't see it being done without a complete rewrite.

I apologize, I should have made it clear that's been discussed in the core committers channel and @Atinux has already started planning a decoupled V3 rewrite as I've described.

Also @ishitatsuyuki -- I understand the frustration, but please don't be offended by the response time from people who have day jobs to attend to. Please bear in mind @Atinux does Nuxt work in his free time and the volume of discussions is just too high for him to be able to pick up on everything, especially something that, in your words, is _trivial and low demand_ :)

So to sum up, the takeaway here is that:

- There are solid plans to decouple as much as possible from the Nuxt Core
- GZip compression is specialized enough that it shouldn't be in Nuxt Core

I'd ask you to concentrate your efforts, if you still feel like engaging, on precisely that: helping us to rearchitect and decouple this functionality from Nuxt Core in the coming months. It will be, like you said, a major rewrite but one that should be extremely rewarding in the end.
